File: src/db/repo/scoped.ts
/**
 * Shared machinery for the repositories.
 *
 * Everything here may import the `db` handle; nothing outside `src/db/**` may
 * (ESLint enforces it). The repositories are the only place a query is written,
 * and every one of their exported functions takes a {@link Scope} first.
 */

import type { AnyColumn, SQL } from 'drizzle-orm';
import { db } from '@/db';
import type { Scope, ScopedRootTable } from '../scope';

/**
 * What a repository function runs its query on: the pool, or an open
 * transaction. Callers inside `src/db/**` pass a transaction when several
 * writes have to land as one unit; everyone else takes the default.
 */
export type DbHandle = typeof db | Parameters<Parameters<typeof db.transaction>[0]>[0];

/**
 * Runs `fn` inside one transaction, handing it the transaction to pass on to the
 * repository functions it calls. The only way a transaction leaves this layer,
 * which is what keeps "several writes are one unit" a data-access concern rather
 * than something every caller could get subtly wrong.
 */
export function withTransaction<T>(fn: (tx: DbHandle) => Promise<T>): Promise<T> {
  return db.transaction((tx) => fn(tx));
}

/**
 * Restricts a child row to parents that are in scope.
 *
 * Child tables (`machine_models`, `tools`, `prompts`, `prompt_toolsets`,
 * `run_results`) never carry `customer_id` themselves — they inherit it through
 * their foreign key. Phase 3: undefined, there is one implicit workspace.
 */
export function scopeThroughParent(
  scope: Scope,
  childFk: AnyColumn,
  parentTable: ScopedRootTable,
  parentId: AnyColumn,
): SQL | undefined {
  void scope;
  void childFk;
  void parentTable;
  void parentId;
  // Phase 5: return inArray(
  //   childFk,
  //   db.select({ id: parentId }).from(parentTable).where(scopeWhere(scope, parentTable)),
  // );
  return undefined;
}

/**
 * Reads a nullable numeric aggregate (`avg`, and any `sum` that can come back
 * empty).
 *
 * `Number(null)` is `0`, so `.mapWith(Number)` on an aggregate over no rows
 * would turn "nothing was measured" into "0 tok/s". Null has to survive. Counts
 * are the opposite case and keep `.mapWith(Number)`: Postgres returns `bigint`
 * as a string, and a count is never null.
 */
export function num(value: unknown): number | null {
  if (value === null || value === undefined) return null;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
}

File: backend/app/repos/scoped.py
"""Shared machinery for the repositories."""

from collections.abc import AsyncIterator
from contextlib import asynccontextmanager
from datetime import UTC, datetime
from typing import Any

from sqlalchemy import ColumnElement, Delete, Select, Update, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import InstrumentedAttribute

from app.scope import Scope, ScopedRoot, scope_where, visible_where


def utc_now() -> datetime:
    """An aware "now".

    Every timestamp column is ``timestamptz`` and server code therefore always
    holds aware datetimes; a naive one would be interpreted in the process's
    local zone on the way in.
    """
    return datetime.now(UTC)


def apply_where[S: (Select[Any], Update, Delete)](
    statement: S, condition: ColumnElement[bool] | None
) -> S:
    """Adds a predicate to a statement, or leaves it alone when there is none.

    ``None`` is what :func:`app.scope.where_scoped` returns for a system scope —
    "every workspace" — and SQLAlchemy refuses a literal ``None`` in
    ``.where()``, so the distinction is made here rather than in every query.
    """
    if condition is None:
        return statement
    return statement.where(condition)


def scope_through_parent(
    scope: Scope,
    child_fk: InstrumentedAttribute[Any],
    parent: ScopedRoot,
    parent_id: InstrumentedAttribute[Any],
    *,
    visible: bool = False,
) -> ColumnElement[bool] | None:
    """Restricts a child row to parents that are in scope.

    Child tables never carry ``customer_id`` themselves — they inherit it
    through their foreign key. A read can express that as a join; an UPDATE or
    DELETE cannot, which is what this subquery is for.

    ``visible=True`` inherits through :func:`~app.scope.visible_where` instead
    of ownership, so a global endpoint's ``endpoint_models`` and a global
    toolset's ``tools`` come along with the parent a workspace can see. It is
    opt-in for the same reason the seam itself is: a child read that forgets it
    shows nothing shared, which is a missing feature and never a leak. A write
    must not pass it — a shared parent's children are still only editable where
    the parent lives.
    """
    parent_scope = visible_where(scope, parent) if visible else scope_where(scope, parent)
    if parent_scope is None:
        return None
    return child_fk.in_(select(parent_id).where(parent_scope))


@asynccontextmanager
async def transaction(session: AsyncSession) -> AsyncIterator[AsyncSession]:
    """Runs a block as one unit of work.

    The only way a transaction leaves this layer, which is what keeps "these
    writes are one unit" a data-access concern rather than something every
    caller could get subtly wrong (run creation freezing a run, its result rows
    and a model sighting together is the case that needs it).

    A session that is already in a transaction gets a SAVEPOINT, so nesting is
    safe — notably under the integration harness, which wraps each test.
    """
    if session.in_transaction():
        async with session.begin_nested():
            yield session
    else:
        async with session.begin():
            yield session
